Linear complementary constraint

hongkai-dai opened the pull request:
We can now solve a problem with linear complementary constraint, and other constraint/cost, through snopt.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6857)
<!-- Reviewable:end -->

hongkai-dai commented:
+@sammy-tri for feature review please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-KriAL3c5WQZ9xe-BSSv:bjquaum)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/mathematical_program.cc, line 87 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkajmaNmGj0CQpL3IY:-KrkajmcUUBuYvBu8Ubw:b-88p691) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/mathematical_program.cc#L87)):*
> ```cpp
> 
> // Solvers for generic systems of constraints and costs.
> AttributesSet kGenericSolverCapabilities =
> ```

Do we now need a `kSnoptSolverCapabilities`which is distinct from `kGenericSolverCapabilites` (because it would also include `kLinearComplementarityConstraint`)? (and then test for it in `MathematicalProgram::Solve`?).  It seems to be like Solve() wouldn't work now unless you called `SnoptSolver` directly, which may also indicate a lack of sufficient testing.

---

*[drake/solvers/mathematical_program.cc, line 504 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkbPC-Ha8kDFhdCSRu:-KrkbPC0ktZhIHq5HBf8:b-mt0k4j) ([raw file](https://github.com/robotlocomotion/drake/blob/022a7fe2bf5ee5425fd881c7b0295a8e9cc30aaa/drake/solvers/mathematical_program.cc#L504)):*
> ```cpp
>   // Linear Complementarity Constraint cannot currently coexist with any
>   // other types of constraint or cost.
>   // (TODO(ggould-tri) relax this to non-overlapping bindings, possibly by
> ```

There's another TODO in `mathematical_program.h` about non-overlapping bindings near the declaration of `linear_complementarity_constraints_`.  It seems strange to remove one of them but not the other?

---

*[drake/solvers/snopt_solver.cc, line 475 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkcS_VdfuMrv-2PaKF:-KrkcS_VdfuMrv-2PaKG:baz754j) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/snopt_solver.cc#L475)):*
> ```cpp
> template <typename C>
> void UpdateLinearConstraint(const MathematicalProgram& prog,
>                        const std::vector<Binding<C>>& linear_constraints,
> ```

odd indentation

---

*[drake/solvers/snopt_solver.cc, line 717 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkdGsumQzvzXKRHUWJ:-KrkdGsumQzvzXKRHUWK:b65w67m) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/snopt_solver.cc#L717)):*
> ```cpp
>     return SolutionResult::kSolutionFound;
>   } else {
>     drake::log()->info("Snopt returns code {}\n", info);
> ```

BTW I have mixed feelings about this being at info log level instead of debug, but since it's an error case which probably isn't going to be found during normal execution, I think I'm OK with it assuming platform review doesn't mind either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/mathematical_program.cc, line 87 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkajmaNmGj0CQpL3IY:-KrlS7sRmjH-28JcgkkU:b-sta43g) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/mathematical_program.cc#L87)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Do we now need a `kSnoptSolverCapabilities`which is distinct from `kGenericSolverCapabilites` (because it would also include `kLinearComplementarityConstraint`)? (and then test for it in `MathematicalProgram::Solve`?).  It seems to be like Solve() wouldn't work now unless you called `SnoptSolver` directly, which may also indicate a lack of sufficient testing.
</blockquote></details>

Done. Good call. I changed one test to use `MathematicalProgram::Solver`.

---

*[drake/solvers/mathematical_program.cc, line 504 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkbPC-Ha8kDFhdCSRu:-KrlSeBg4L5oltlF0MkS:b-w4e0cr) ([raw file](https://github.com/robotlocomotion/drake/blob/022a7fe2bf5ee5425fd881c7b0295a8e9cc30aaa/drake/solvers/mathematical_program.cc#L504)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

There's another TODO in `mathematical_program.h` about non-overlapping bindings near the declaration of `linear_complementarity_constraints_`.  It seems strange to remove one of them but not the other?
</blockquote></details>

Done. Good catch. I changed the documentation in `mathematical_program.h`.

---

*[drake/solvers/snopt_solver.cc, line 475 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkcS_VdfuMrv-2PaKF:-KrlTm2gqxLILSk3bE31:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/snopt_solver.cc#L475)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

odd indentation
</blockquote></details>

Done.

---

*[drake/solvers/snopt_solver.cc, line 717 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrkdGsumQzvzXKRHUWJ:-KrlXPb2t-5cBHw4BN_p:bd7bx20) ([raw file](https://github.com/robotlocomotion/drake/blob/c32ac264df163d278c144428348ed34cdcad605c/drake/solvers/snopt_solver.cc#L717)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW I have mixed feelings about this being at info log level instead of debug, but since it's an error case which probably isn't going to be found during normal execution, I think I'm OK with it assuming platform review doesn't mind either.
</blockquote></details>

Done. I changed to debug level. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (with one minor nit) +@soonho-tri for platform review

+(status: curate commits before merging) 

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 2225 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krl_cz75lQIZGTN91wA:-Krl_cz75lQIZGTN91wB:b-5rs6gw) ([raw file](https://github.com/robotlocomotion/drake/blob/96d8aa68722a415a765eb73ff93b61c0fe8277c3/drake/solvers/mathematical_program.h#L2225)):*
> ```objc
> 
>   // Invariant:  The bindings in this list must be non-overlapping, when calling
>   // Linear Complementarity solver like Moby. If this constrained is solved
> ```

BTW typo "constraint"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-Krl_WyQWJZNg0Zp8ljD:b-qb5951)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@hongkai-dai , could you check the CI failures?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-Krlm0IvuliBfZVn_RFE:bp9fbts)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@soonho-tri , the CI failure should be fixed after running cpplint.

---

Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 2225 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krl_cz75lQIZGTN91wA:-KrlyEzBn04UUKBEefnD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/96d8aa68722a415a765eb73ff93b61c0fe8277c3/drake/solvers/mathematical_program.h#L2225)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW typo "constraint"?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-KrlyGaD-Jb0h0wg9ptO:b-3pxn5b)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 5 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm3FnsLHLk7ySry6k0:b-vpyuhw) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
> ```CMake
> endif()
> 
> drake_add_cc_test(complementary_problem_test)
> ```

BTW, we've removed snopt from cmake build in https://github.com/RobotLocomotion/drake/commit/548d7c5ea38aadc610584b9d6e0495345b50a252. I guess that means we do not need to include this test in CMake.

@jamiesnape , could you confirm it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm5xz5yCamuLUXG6Ov:b-1im6l3) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, we've removed snopt from cmake build in https://github.com/RobotLocomotion/drake/commit/548d7c5ea38aadc610584b9d6e0495345b50a252. I guess that means we do not need to include this test in CMake.

@jamiesnape , could you confirm it?
</blockquote></details>

Yes, SNOPT is gone. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm6G2T1wdyubJ_k8Vu:bfkn05u) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Yes, SNOPT is gone. 
</blockquote></details>

No real need to include the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm6UXmPSBfoZP4BHRB:b-pavffu) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

No real need to include the test.
</blockquote></details>

@jamiesnape , thanks for the check. @hongkai-dai , let's remove it from this file then.

@jamiesnape , I'm wondering why there was no test failure from CMake builds then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm6jJA73FoRljRBoNu:b4pfupv) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@jamiesnape , thanks for the check. @hongkai-dai , let's remove it from this file then.

@jamiesnape , I'm wondering why there was no test failure from CMake builds then.
</blockquote></details>

The test checks whether SNOPT is available, receives the answer "no" and skips everything in the conditional.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm6w1oNkwUf4WCbGEl:b569cij) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

The test checks whether SNOPT is available, receives the answer "no" and skips everything in the conditional.
</blockquote></details>

The same reason that it works on Bazel non-SNOPT builds. It could have a `"snopt"` tag if you want to save a few seconds of CI time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-Krm7BAs6IVMXOi9n5-6:b-i4eglb) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
@jamiesnape , thank you.

> It could have a "snopt" tag if you want to save a few seconds of CI time.

@hongkai-dai , could you handle this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-KrmDKkX9-kY-oXSRK9f:b-u0liv3) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@jamiesnape , thank you.

> It could have a "snopt" tag if you want to save a few seconds of CI time.

@hongkai-dai , could you handle this?
</blockquote></details>

I removed the test from cmake.

Unlike `gurobi_test_tags()` defined in `tools/gurobi.bzl` or `mosek_test_tags()` defined in `tools/mosek.bzl`, there is not a `tools/snopt.bzl` file, and `snopt_test_tags` is not defined.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-KrpRsjUkHaGaT5dWMTg:b-ptlwe8) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I removed the test from cmake.

Unlike `gurobi_test_tags()` defined in `tools/gurobi.bzl` or `mosek_test_tags()` defined in `tools/mosek.bzl`, there is not a `tools/snopt.bzl` file, and `snopt_test_tags` is not defined.
</blockquote></details>

The tag is simply `snopt`, e.g., https://github.com/RobotLocomotion/drake/blob/97c64fe0c0ae5e003d40f9f7e685e455ae79dc6a/drake/manipulation/planner/BUILD#L40

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 4 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/test/complementary_problem_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrpnWScFRwbIf_FjJAB:-KrpnWScFRwbIf_FjJAC:b-sfihkb) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L37)):*
> ```cpp
>   Eigen::Matrix<double, 5, 5> M;
>   // clang-format off
> M << 3, -1, 0, 0, 0,
> ```

BTW, nit, the indentation do not look good. Please consider using the following:
```c++
  M << 3,  -1, 0, 0, 0,
      -1, 0.5, 0, 0, 0,
      -1,  -1, 0, 0, 0,
       0,   0, 0, 0, 0,
       0,   0, 0, 0, 0;
``` 

---

*[drake/solvers/test/complementary_problem_test.cc, line 55 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krpnf74KPuOkUj0ifo3:-Krpnf74KPuOkUj0ifo4:bjdc1vv) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L55)):*
> ```cpp
>     EXPECT_NEAR(x_val(0), 1, 1E-6);
>     EXPECT_NEAR(y_val(0), 0, 1E-6);
> }
> ```

BTW, nit, I think there should be two spaces in front.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/solvers/snopt_solver.cc, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krq3b_ZKbvlBRWiIAzN:-KrqHPGD2GggwtFjObgd:b-433ptx) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L24)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, I found that `clang-format off` does not work well with c-style comments. Please consider changing them to the following:

```c++
// clang-format off
#include "snopt.hh"
#include "snfilewrapper.hh"
// clang-format on

</blockquote></details>

Done. Good to know!

---

*[drake/solvers/snopt_solver.cc, line 211 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrpzY32v7RlKvVgCyGC:-KrqI-IJeLRabfpyrg6a:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L211)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Could you add a sentence/paragraph explaining why we need these specializations (instead of relying on `.num_constraint()` and `.Eval()`)? I guess 
</blockquote></details>

Done.

---

*[drake/solvers/snopt_solver.cc, line 362 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krq1aqSAz9SAyHoEmNl:-KrqIIoECA62X-2qVxC2:b-47i3xy) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L362)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I'm afraid that this documentation is not a proper description of what this function does. Maybe we can copy the one above and modify for this specialization?
</blockquote></details>

Done. I explained what this function does. 

I actually think we should point out why LinearComplementaryConstraint only adds one row of constraint, so I keep the formulation `xᵀ(Mx+q) = 0` in the documentation.

---

*[drake/solvers/snopt_solver.cc, line 441 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krq2HFI7e2T_3oFXZ_g:-KrqJVRkZi3vICBMGl2g:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L441)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think this documentation is misleading.
</blockquote></details>

Done.

---

*[drake/solvers/snopt_solver.cc, line 489 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krq3wx1n4SdV3ybkZAV:-KrqJiCGRrIjcOD_XJGj:b-yh16l6) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L489)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, the following works (and should be slightly more efficient):

```c++
        tripletList->emplace_back(
            *linear_constraint_index + it.row(),
            prog.FindDecisionVariableIndex(binding.variables()(k)), it.value());
```
</blockquote></details>

Done. Good call.

---

*[drake/solvers/snopt_solver.cc, line 581 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krq4BdE1y_gu32mn6fR:-KrqJzOTi2dZCUmtOWeI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/snopt_solver.cc#L581)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Again, I think this is not what we do in the code?
</blockquote></details>

Done.

---

*[drake/solvers/test/complementary_problem_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrpnWScFRwbIf_FjJAB:-KrqKI_C8McsSojMB6TD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L37)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, the indentation do not look good. Please consider using the following:
```c++
  M << 3,  -1, 0, 0, 0,
      -1, 0.5, 0, 0, 0,
      -1,  -1, 0, 0, 0,
       0,   0, 0, 0, 0,
       0,   0, 0, 0, 0;
``` 
</blockquote></details>

Done.

---

*[drake/solvers/test/complementary_problem_test.cc, line 55 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krpnf74KPuOkUj0ifo3:-KrqKOCDOCdUwudalB3d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L55)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, I think there should be two spaces in front.
</blockquote></details>

Done.

---

*[drake/solvers/test/CMakeLists.txt, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-Krm3FnsLHLk7ySry6k-:-KrqKYRpuxkuJ4IMAnp0:b-975ghh) ([raw file](https://github.com/robotlocomotion/drake/blob/8b452bf611ee7123d72dddd70e6382cb340c49cf/drake/solvers/test/CMakeLists.txt#L66)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

The tag is simply `snopt`, e.g., https://github.com/RobotLocomotion/drake/blob/97c64fe0c0ae5e003d40f9f7e685e455ae79dc6a/drake/manipulation/planner/BUILD#L40
</blockquote></details>

Done. Thanks! I didn't know we have snopt tag.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (aside minor comments)

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/snopt_solver.cc, line 207 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrqNIu4DultmX5y1gl_:-KrqNIu4DultmX5y1gla:beop843) ([raw file](https://github.com/robotlocomotion/drake/blob/6705e9819383e1d5a670fba13ed2dadd436437ac/drake/solvers/snopt_solver.cc#L207)):*
> ```cpp
> // of the constraint directly. For some other constraint, such as
> // LinearComplementaryConstraint, we will evaluate its nonlinear constraint
> // differently, than its Eval function.
> ```

For my education, could you explain why we compute `Mx + q` in `Eval` function while we compute `xᵀ(Mx + q)` here? It would be nice to add that to the documentation if you think it's worth.

---

*[drake/solvers/test/complementary_problem_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrpnWScFRwbIf_FjJAB:-KrqN-nRLJNqU_-_oZav:b-5bwbeq) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L37)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I should've been more explicit. I wanted to point out the indentation of `M`. We need two spaces in front...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-KrqNWGWx31QMHoI7XT5:b-6v8yxj)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
I think the failure in `linux-xenial-clang-ninja-experimental-open-source` is a glitch.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-:-KrqOttLBAyG9Fbsubf-:byr52bt)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/snopt_solver.cc, line 207 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrqNIu4DultmX5y1gl_:-KrqO0ACKFTvqb3gOOzv:burcdl2) ([raw file](https://github.com/robotlocomotion/drake/blob/6705e9819383e1d5a670fba13ed2dadd436437ac/drake/solvers/snopt_solver.cc#L207)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

For my education, could you explain why we compute `Mx + q` in `Eval` function while we compute `xᵀ(Mx + q)` here? It would be nice to add that to the documentation if you think it's worth.
</blockquote></details>

I actually do not know why we implemented `Mx + q` in the `Eval` function. I don't think it is called anywhere. @edrumwri , could you educate us why you chose `Mx + q` in the `Eval` function?

Anyway, it is not right to put `xᵀ(Mx + q)` in the `Eval` function either. There are two reasons
1. When `lower_bound <= Eval(x) <= upper_bound`, it should guarantee that `x` satisfies the constraint. Here `xᵀ(Mx + q)=0` alone does not guarantee that the constraint is satisfied. We also need `x >= 0` and `Mx + q >= 0`.
2. If we instead put `x >= 0, Mx + q >= 0, xᵀ(Mx + q)=0` all into the `Eval` function, that guarantees `x` satisfies the constraint, when `lower_bound <= Eval(x) <= upper_bound`. But in SNOPT, when we add the `LinearComplementaryConstraint` into the problem, we should not blindly add them all as nonlinear constraints. If we do so, we lose the property, that `x>=0` is a BoundingBoxConstraint, and `Mx + q>=0` is a linear constraint. This property is crucial when using SQP solver, like SNOPT. This is explained in the paper "Local Convergence of SQP Methods for Mathematical Programs with Equilibrium Constraints". One of the trick to improve the convergence, is to guarantee that `x>=0` and `Mx+q>=0` for each QP iteration. To achieve this, they need to be imposed as bounding box constraint, and linear constraint respectively, but not as nonlinear constraints, since SQP solver does not guarantee to satisfy nonlinear constraint, in each iteration.

---

*[drake/solvers/test/complementary_problem_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrpnWScFRwbIf_FjJAB:-KrqQSu30OKjrLOlwVyp:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/cabc2d31fd34313e12bacc65df0b3f1d1a644608/drake/solvers/test/complementary_problem_test.cc#L37)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I should've been more explicit. I wanted to point out the indentation of `M`. We need two spaces in front...
</blockquote></details>

Done. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/snopt_solver.cc, line 207 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrqNIu4DultmX5y1gl_:-KrqTSiT-MrYqguVHA5z:bj05rgq) ([raw file](https://github.com/robotlocomotion/drake/blob/6705e9819383e1d5a670fba13ed2dadd436437ac/drake/solvers/snopt_solver.cc#L207)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I actually do not know why we implemented `Mx + q` in the `Eval` function. I don't think it is called anywhere. @edrumwri , could you educate us why you chose `Mx + q` in the `Eval` function?

Anyway, it is not right to put `xᵀ(Mx + q)` in the `Eval` function either. There are two reasons
1. When `lower_bound <= Eval(x) <= upper_bound`, it should guarantee that `x` satisfies the constraint. Here `xᵀ(Mx + q)=0` alone does not guarantee that the constraint is satisfied. We also need `x >= 0` and `Mx + q >= 0`.
2. If we instead put `x >= 0, Mx + q >= 0, xᵀ(Mx + q)=0` all into the `Eval` function, that guarantees `x` satisfies the constraint, when `lower_bound <= Eval(x) <= upper_bound`. But in SNOPT, when we add the `LinearComplementaryConstraint` into the problem, we should not blindly add them all as nonlinear constraints. If we do so, we lose the property, that `x>=0` is a BoundingBoxConstraint, and `Mx + q>=0` is a linear constraint. This property is crucial when using SQP solver, like SNOPT. This is explained in the paper "Local Convergence of SQP Methods for Mathematical Programs with Equilibrium Constraints". One of the trick to improve the convergence, is to guarantee that `x>=0` and `Mx+q>=0` for each QP iteration. To achieve this, they need to be imposed as bounding box constraint, and linear constraint respectively, but not as nonlinear constraints, since SQP solver does not guarantee to satisfy nonlinear constraint, in each iteration.
</blockquote></details>

@hongkai-dai , thanks for the explanation! I mark it as satisfied.

BTW/FYI, I've found that there is `LCP::DoCheckSatisfied` which checks the three conditions completely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/solvers/snopt_solver.cc, line 207 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6857#-KrqNIu4DultmX5y1gl_:-KrqYUoHmkdCfbAnLREF:brq0ouc) ([raw file](https://github.com/robotlocomotion/drake/blob/6705e9819383e1d5a670fba13ed2dadd436437ac/drake/solvers/snopt_solver.cc#L207)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@hongkai-dai , thanks for the explanation! I mark it as satisfied.

BTW/FYI, I've found that there is `LCP::DoCheckSatisfied` which checks the three conditions completely.
</blockquote></details>

Thanks! Good to know about `DoCheckSatisfied`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6857)*
<!-- Sent from Reviewable.io -->
